### ISSUE TITLE
vim.list_extend vs. vim.tbl_extend

### ISSUE THREAD

**Issue by Julian** [neovim/neovim#14544]
`vim.list_extend` and `vim.tbl_extend` are inconsistent -- the latter returns a copy, while the former mutates its argument.

Is it too late to "fix" / change that? I'd love a `list_extend` that returns a copy instead, it seems to be a common need for me when writing Lua code, and then if versions with mutation are desirable presumably some second mutating API for both?

As usual feel free to close if this is a ridiculous idea.

**Comment by folke** [comment]
There's definitely code in the wild (including one of my own plugins) that expects `vim.list_extend` to extend the given list with the additional arguments.

A [search](https://github.com/search?q=vim.list_extend&type=code) on Github shows a lot of usage of `list_extend` on how it was intended to be used.

Changing this would potentially break a lot of people's code.

**Comment by Julian** [comment]
Yep, if replacing them is too risky, we could instead obviously introduce new APIs that work the same as each other and deprecate these.

**Comment by JosefLitos** [comment]
I was just looking around how to append a table to a table and didn't know about this function first. And according to a [comment](https://stackoverflow.com/questions/1410862/concatenation-of-tables-in-lua#comment46607955_15278426) on that matter it turns out using `table.insert` is unnecessarily inefficient. Could this then use the simple lua `dst[#dst+1]=src[i]`?

**Comment by justinmk** [comment]
Related: https://github.com/neovim/neovim/issues/24572
